### ISSUE TITLE
caml_release_runtime_and_perform_effect

### ISSUE THREAD

**Issue by craff** [ocaml/ocaml#11822]
Hello,

When calling C function that may block or run a long time, it would be nice to perform an effect 
and release the runtime at the same time. This way, an OCaml scheduler can take over and let the C operation run in parallel.

The continuation in the effect should be garded by some "condition" and be only usable when the C code is asking
to acquire the runtime again.

**Comment by kayceesrk** [comment]
> it would be nice to perform an effect and release the runtime at the same time. This way, an OCaml scheduler can take over and let the C operation run in parallel.

The proposed operation creates a thread of execution which runs in parallel with the computation that performs the effect. This is not how effect handlers work. Effect handlers only let you suspend the current computation. The proposed operation works like a thread `spawn`/`async`. One would need a corresponding `join`/`await`. Unclear how to reconcile this with the current semantics.

In the past, there was a proposal to allow effects to be performed from C [1] using the idea from Threads Yield Continuations [2] paper. Since then, there have been several developments supporting effect handlers in C [3,4]. Studying these should give us a better idea of deciding how effect handlers _should_ work when effects are performed from C. 

[1] https://github.com/ocaml-multicore/ocaml-multicore/issues/146
[2] https://www.cs.indiana.edu/~dyb/pubs/LaSC-10-3-pp223-236.pdf
[3] https://github.com/koka-lang/libmprompt
[4] https://dl.acm.org/doi/abs/10.1145/3563445

**Comment by craff** [comment]
Thanks, you are right, this would require a new thread. And it may be too much.

Still, with ocaml 5, I am not sure when to release the runtime. Exemple SSL_read, on a non blocking socket. 
It can still take some time. 
- not releasing the runtime could delay a stop the world phase
- releasing it requires to write in a temporary buffer and copy that buffer to ocaml memory after reacquiring the runtime.

Really wonder what is better ?

**Comment by craff** [comment]
I just made some tests with ocaml-ssl, in very stressing condition. No significative differences.

**Comment by kayceesrk** [comment]
> not releasing the runtime could delay a stop the world phase

You don't need to release the runtime for the stop-the-world phase. The GC can run at any safe points (allocation points, poll points, etc), just like OCaml 4. You shouldn't have to change anything in your program compared to OCaml 4 to accommodate the stop-the-world GC.

**Comment by kayceesrk** [comment]
> I just made some tests with ocaml-ssl, in very stressing condition. No significative differences.

I assume that the program is a sequential program? OCaml 5 was explicitly designed to be backwards compatible in terms of features and performance (to some extent). It is expected that you will not see significant differences on sequential programs.

**Comment by gadmm** [comment]
> You don't need to release the runtime for the stop-the-world phase. The GC can run at any safe points (allocation points, poll points, etc), just like OCaml 4. You shouldn't have to change anything in your program compared to OCaml 4 to accommodate the stop-the-world GC.

IIUC the wish is to call a C function that could run a long time and not regularly poll (e.g. because it is foreign or because it blocks) and in this case one absolutely wants to release the domain lock.

You mention general support of doing effects from C code in general, but as an alternative one could offload a computation to a systhread by hand and have some helper functions for doing just that (for a specific OCaml-side scheduler). It might be worth looking at the design of similar things for Lwt, e.g. [Mwt](https://github.com/hcarty/mwt) and [Lwt_preemptive](https://ocsigen.org/lwt/5.3.0/api/Lwt_preemptive).

**Comment by dbuenzli** [comment]
Incidentally I was thinking about the same thing the other day (though in my use case I managed to convince myself there was likely little benefit over continuing with the same thread). 

At the risk of paraphrasing @gadmm, @craff did you consider using an OCaml thread pool ? Basically perform an effect to make the operation, your handler/scheduler throws the operation at the thread pool, the chosen thread calls C, releases the runtime, reacquires it and deposits the result in a data structure where your cooperative handler/scheduler can pick up the result to invoke the continuation with.

**Comment by kayceesrk** [comment]
Thanks for the clarification. Offloading the computation to a different systhread on the same domain seems like the most useful thing to do.

**Comment by craff** [comment]
Thanks for your comments. 

About thread: a fresh thread would be too much for a non-blocking sslRead and sslWrite. I had indeed a thread pool in mind, because I already have a domain pool doing IO on communication socket, with a scheduler and cooperative concurrency on each domain. I could keep a few domains for non IO task like sslRead and sslWrite in the hope to have less latency on other socket IO.

From running the tests in [simple_httpd](https://github.com/craff/simple_httpd) I am not sure it is worth it. 

Probably a pool of C threads doing the C job would be a better idea. I really need to measure the average timing of sslRead and sslWrite in non blocking mode first.
 
About runtime lock: using a branch I did of [ocaml-ssl](https://github.com/savonet/ocaml-ssl/pull/101) with more alternatives functions that do not release the runtime for sslRead, sslWrite, sslAccept, and other frequently called functions, I do see a small improvement under heavy load. So it seems that not releasing the lock during the C call does
not deteriorate the performance.

Any way even as it is, it works very well! Thanks for OCaml 5, its domain and effect, I had fun writing this and it seems very robust.
 
By the way, this issue is more a discussion than an issue now, feel free to close it!
